### PR TITLE
sensor: drivers: Panasonic sngcja5 particle matter sensor

### PR DESCRIPTION
**Pull request by pideu-sj**
Support for Panasonic's [SN-GCJA5](https://industry.panasonic.eu/products/components/sensors/particulate-matter-sensor) Particle Matter Sensor.
Added sensor drivers and sensor sample for example usage.

### CODE REVIEW

File: drivers/sensor/sngcja5/sngcja5.c
Comment (pideu-sj) [inline drivers/sensor/sngcja5/sngcja5.c:163]: I'm always getting following error message in the twister runs: 
`zephyr/drivers/sensor/sngcja5/sngcja5.c:163:39: error: 'sngcja5_driver_api' defined but not used [-Werror=unused-const-variable=]`

Can someone explain me what I'm doing wrong, or what I have to do to pass?

File: drivers/sensor/sngcja5/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/sngcja5/Kconfig:9]:
```suggestion
	depends on DT_HAS_PANASONIC_SNGCJA5_ENABLED`
	select I2C
```

File: drivers/sensor/sngcja5/sngcja5.c
Comment (MaureenHelm) [inline drivers/sensor/sngcja5/sngcja5.c:35]: Just use `i2c_burst_read_dt` instead

File: drivers/sensor/sngcja5/sngcja5.c
Comment (MaureenHelm) [inline drivers/sensor/sngcja5/sngcja5.c:149]: `i2c_is_ready_dt`

File: drivers/sensor/sngcja5/sngcja5.c
Comment (MaureenHelm) [reply drivers/sensor/sngcja5/sngcja5.c:163]: There probably aren't any enabled instances of `panasonic,sngcja5` in your devicetree

File: dts/bindings/sensor/panasonic,sngcja5.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/panasonic,sngcja5.yaml:8]:
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

File: drivers/sensor/sngcja5/sngcja5.c
Comment (MaureenHelm) [inline drivers/sensor/sngcja5/sngcja5.c:115]: Need to use a more specific channel type. If there isn't one in `enum sensor_channel` that makes sense for this sensor, then consider defining a driver specific channel. See drivers that use `SENSOR_CHAN_PRIV_START` for example

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/sngcja5/Kconfig
Comment (pideu-sj) [reply drivers/sensor/sngcja5/Kconfig]: Pushed as requested in latest pull request update.

File: drivers/sensor/sngcja5/sngcja5.c
Comment (pideu-sj) [reply drivers/sensor/sngcja5/sngcja5.c]: Pushed as requested in latest pull request update.

File: drivers/sensor/sngcja5/sngcja5.c
Comment (pideu-sj) [reply drivers/sensor/sngcja5/sngcja5.c]: Pushed as requested in latest pull request update.

File: drivers/sensor/sngcja5/sngcja5.c
Comment (pideu-sj) [reply drivers/sensor/sngcja5/sngcja5.c:155]: But I've added the _compatible = "panasonic,sngcja5";_ to my _arduino_i2c.overlay_ file. So i thought that it has to be in the devicetree afterwards?

Comment (MaureenHelm) [reply drivers/sensor/sngcja5/sngcja5.c:155]: is `status = okay` ? do you see the node in the build artifact `zephyr.dts`? if you don't, that would suggest you're not including the overlay in the build

File: samples/sensor/sngcja5/src/main.c
Comment (MaureenHelm) [inline samples/sensor/sngcja5/src/main.c:20]: Please make this sample more generic so it can support other particle sensor drivers. See `samples/sensor/accel_polling` for example

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by pideu-sj** [comment]
Hello @MaureenHelm 
sorry was caught on other projects on last weeks, but will go on here in the next days. Could you please remove the stale label?

File: drivers/sensor/sngcja5/sngcja5.c
Comment (pideu-sj) [reply drivers/sensor/sngcja5/sngcja5.c:48]: When using `sys_be16_to_cpu` like in following snippet:
```
static int read_register_2(const struct i2c_dt_spec *dev, uint8_t addr, uint16_t *value)
{
	if (i2c_burst_read_dt(dev, addr, (uint8_t *)value, sizeof(*value)) < 0) {
		LOG_ERR("i2c_burst_read_dt() @ i2c failed");
		return -EIO;
	}

	*value = sys_be16_to_cpu(*value);

	return 0;
}
```
Like it's done for example in `drivers/sensor/tmp112/tmp112.c`, i would get the wrong byte order for my read data. Should be: 
``
new sample:
pc0_5.1: 0x0003
``
But is:
``
new sample:
pc0_5.1: 0x0300
``
So i guess i have to stick to the previous solution.

File: samples/sensor/sngcja5/src/main.c
Comment (pideu-sj) [reply samples/sensor/sngcja5/src/main.c:20]: Did that on last pull request update.

File: drivers/sensor/sngcja5/sngcja5.c
Comment (pideu-sj) [reply drivers/sensor/sngcja5/sngcja5.c:29]: See https://github.com/zephyrproject-rtos/zephyr/pull/55765/commits/755d6357018471d9dfe3ae15d5a7d58dba58bd12#r1212827727

File: include/zephyr/drivers/sensor.h
Comment (pideu-sj) [reply include/zephyr/drivers/sensor.h]: Updated comment

File: include/zephyr/drivers/sensor.h
Comment (pideu-sj) [reply include/zephyr/drivers/sensor.h]: Did this on last pull-request update.

File: drivers/sensor/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/Kconfig:182]: Looks like there were some rebasing problems

Comment (pideu-sj) [reply drivers/sensor/Kconfig]: Oh Yeah. Not sure how that happened. Fixed in pull-request update. Thanks!

File: drivers/sensor/CMakeLists.txt
Comment (pideu-sj) [reply drivers/sensor/CMakeLists.txt]: You are right. Again not sure, how that made it into the commit. Fixed that in newest update. Thanks.

File: drivers/sensor/sngcja5/sngcja5.c
Comment (pideu-sj) [reply drivers/sensor/sngcja5/sngcja5.c:149]: Yes I see the node in the build artifact zephyr.dts. Don't getting this error message in the twister runs anymore.

**Review by MaureenHelm — COMMENTED** [review]
Please rename the sample and update the documentation to reflect the change making it generic to any particle sensor driver

File: include/zephyr/drivers/sensor.h
Comment (pideu-sj) [reply include/zephyr/drivers/sensor.h:118]: I've found sensors like the `DFRobot SEN0460` that has the same 6 values available (https://wiki.dfrobot.com/Gravity_PM2.5_Air_Quality_Sensor_SKU_SEN0460). But have also found sensor like the `SPS30` that just provides 5 count values (https://sensirion.com/media/documents/8600FF88/616542B5/Sensirion_PM_Sensors_Datasheet_SPS30.pdf).

**Comment by pideu-sj** [comment]
> Please rename the sample and update the documentation to reflect the change making it generic to any particle sensor driver

Done in last update. Changed naming and documentation from sample `sngcja5` to `particulate_matter`.

**Comment by pideu-sj** [comment]
@MaureenHelm 

Did you already found some time to review?

**Comment by pideu-sj** [comment]
Hi @galak @avisconti @teburd , 

can maybe someone else review?

File: include/zephyr/drivers/sensor.h
Comment (avisconti) [reply include/zephyr/drivers/sensor.h:118]: shouldn't this change be in a separate commit?

**Review by avisconti — DISMISSED** [review]
LGTM (with a small non blocking comment, though)

**Comment by pideu-sj** [comment]
@MaureenHelm 
could we maybe reopen this PR and remove the stale label?

It's been a while, but in comment https://github.com/zephyrproject-rtos/zephyr/pull/55765#issuecomment-1579962884 I said, that last requested changes from you site are added. So can we merge this PR now? @avisconti already gave the approval in the meantime.

**Review by teburd — DISMISSED** [review]

**Comment by pideu-sj** [comment]
Hi @teburd and @avisconti
had to rebase to match the compliance check and update manifest to pass all checks after removal of stale label.

**Review by teburd — DISMISSED** [review]

**Comment by pideu-sj** [comment]
Hi @MaureenHelm, @avisconti,
could you maybe find some minuets to check if we can merge the PR?
Thanks a lot.

**Comment by pideu-sj** [comment]
Hi to all listed reviewers,
if @MaureenHelm and @avisconti are not available, could someone else maybe review?
Thanks a lot!

**Comment by pideu-sj** [comment]
Hello to the reviewers in the list @avisconti, @nashif, @galak, @yperess, @kartben, @tristan-google,
the Pull-Request already got two times the stale label. Can maybe someone else review if @MaureenHelm is not available?
Thanks a lot!

File: drivers/sensor/sngcja5/sngcja5.c
Comment (kartben) [reply drivers/sensor/sngcja5/sngcja5.c:43]: These are actually little-endian, not big-endian, so using `*_le*.` versions of the conversion macros should work as expected

File: include/zephyr/drivers/sensor.h
Comment (kartben) [reply include/zephyr/drivers/sensor.h:119]: At a minimum this should be a different commit, yes. 

But I am -1 on adding this new "array" channel as both your linked datasheets actually show that different sensors will likely use different "classes" of particle counts in this array i.e PM 0.5/1.0/2.5/4.0/10.0 for SPS30, and PM 0.3/0.5/1.0/2.5/5.0/10.0, none of which correspond to the ranges proposed here. HM3300/3600 sensors have yet another range (https://files.seeedstudio.com/wiki/Grove-Laser_PM2.5_Sensor-HM3301/res/HM-3300%263600_V2.1.pdf) and so on.

What would probably work though is adding `SENSOR_CHAN_PC_{X_Y}` channels.

**Review by kartben — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/sensor.h
Comment (pideu-sj) [reply include/zephyr/drivers/sensor.h:119]: @kartben 
Thanks a lot for taking over and reviewing.

On first step I spent an own commit for this change, like you've requested.

And you are right, it's difficult, to generalize the particle count values for several sensors.

> What would probably work though is adding SENSOR_CHAN_PC_{X_Y} channels.

Could you give an example, how this would look like for my sngcja5 sensor with the PM 0.5/1.0/2.5/5.0/7.5/10.0 ? Haven't seen it in that way here before?

Comment (kartben) [reply include/zephyr/drivers/sensor.h:119]: > > What would probably work though is adding SENSOR_CHAN_PC_{X_Y} channels.
> 
> Could you give an example, how this would look like for my sngcja5 sensor with the PM 0.5/1.0/2.5/5.0/7.5/10.0 ? Haven't seen it in that way here before?

I mean just like you currently have `SENSOR_CHAN_PM_1_0`, `SENSOR_CHAN_PM_2_5`, etc. there could also be new channels for `SENSOR_CHAN_PC_0_3`, `SENSOR_CHAN_PC_0_5`, etc.?
Note that something like #63830 would probably make things easier here, but in the meantime I think adding channels for the common particle count sizes would work, but I will let @MaureenHelm and other core contributors to the sensor API confirm.

**Comment by pideu-sj** [comment]
@MaureenHelm and other core contributors to the sensor API,
could you confirm, that I should change sensor channels for particle count, like @kartben suggested here: https://github.com/zephyrproject-rtos/zephyr/pull/55765/files#r1391162757
Thanks a lot.

File: include/zephyr/drivers/sensor.h
Comment (teburd) [reply include/zephyr/drivers/sensor.h:119]: Please go with this route today

Comment (yperess) [reply include/zephyr/drivers/sensor.h:119]: I'm not really seeing the need for this to be a part of sensor.h, why can't this highly specific 6 value channel be added in sngcja5.h? I would also like to add that with the new async API you're able to have an index along with a channel which would service this model much better. SENSOR_CHAN_PARTICLE_COUNT would be used along with an index to provide any number of dimensions to the measurement. We would still need a way to describe what each index means.

File: samples/sensor/particulate_matter/sample.yaml
Comment (yperess) [inline samples/sensor/particulate_matter/sample.yaml:1]: I'm going to let @MaureenHelm decide the path on this, but I'm not seeing a strong benefit to adding yet another sensor specific sample to sensors. The sensor shell sample can do this just fine.

**Review by yperess — CHANGES_REQUESTED** [review]

File: samples/sensor/particulate_matter/sample.yaml
Comment (pideu-sj) [reply samples/sensor/particulate_matter/sample.yaml:1]: @yperess, thanks for reviewing. 
Okay, so I will wait for decision of @MaureenHelm, before I change this in the driver.

**Pull request closed without merging**
